cannon/serialize/binary.go

```diff
@@ -0,0 +1,75 @@
+package serialize
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"reflect"
+
+	"github.com/ethereum-optimism/optimism/op-service/ioutil"
+)
+
+// Serializable defines functionality for a type that may be serialized to raw bytes.
+type Serializable interface {
+	// Serialize encodes the type as raw bytes.
+	Serialize(out io.Writer) error
+
+	// Deserialize decodes raw bytes into the type.
+	Deserialize(in io.Reader) error
+}
+
+func LoadSerializedBinary[X any](inputPath string) (*X, error) {
+	if inputPath == "" {
+		return nil, errors.New("no path specified")
+	}
+	var f io.ReadCloser
+	f, err := ioutil.OpenDecompressed(inputPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open file %q: %w", inputPath, err)
+	}
+	defer f.Close()
+
+	var x X
+	serializable, ok := reflect.ValueOf(&x).Interface().(Serializable)
+	if !ok {
+		return nil, fmt.Errorf("%T is not a Serializable", x)
+	}
+	err = serializable.Deserialize(f)
+	if err != nil {
+		return nil, err
+	}
+	return &x, nil
+}
+
+func WriteSerializedBinary(outputPath string, value Serializable, perm os.FileMode) error {
+	if outputPath == "" {
+		return nil
```

Comment:
It matches the WriteJSON behaviour so that nothing is written if the path isn't specified. It is a bit weird, but given we interpret `-` as meaning output to system out here I think it's probably reasonable. It's not obvious from the method name that it will do that kind of handling though.

https://github.com/ethereum-optimism/optimism/blob/98e1cf267aa149de21ee82aa36950d4adcd5f9ce/op-service/jsonutil/json.go#L36-L38

Logged https://github.com/ethereum-optimism/optimism/issues/11775 to look into how to clean this up.